stl/inc/memory

```diff
@@ -3150,26 +3171,269 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
+        const auto _Rep = this->_Lock_and_load();
+        _Result._Ptr    = this->_Ptr;
+        _Result._Rep    = _Rep;
+        _Result._Incref();
+        this->_Store_and_unlock(_Rep);
+        return _Result;
+    }
+
+    operator shared_ptr<_Ty>() const noexcept {
+        return load();
+    }
+
+    shared_ptr<_Ty> exchange(shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);
+        shared_ptr<_Ty> _Result;
+        _Result._Rep = this->_Lock_and_load();
+        _Result._Ptr = this->_Ptr;
+        this->_Ptr   = _Value._Ptr;
+        this->_Store_and_unlock(_Value._Rep);
+        _Value._Ptr = nullptr; // ownership of _Value ref has been given to this, silence decrement
+        _Value._Rep = nullptr;
+        return _Result;
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Order);
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);
+        auto _Rep = this->_Lock_and_load();
+        if (this->_Ptr == _Expected._Ptr && _Rep == _Expected._Rep) {
+            _STD swap(this->_Ptr, _Desired._Ptr);
+            _STD swap(_Rep, _Desired._Rep);
+            this->_Store_and_unlock(_Rep);
+            return true;
+        }
+        _Ref_count_base* _Expected_rep = _Expected._Rep;
+        _Expected._Ptr                 = this->_Ptr;
+        _Expected._Rep                 = _Rep;
+        _Expected._Incref();
+        this->_Store_and_unlock(_Rep);
+        if (_Expected_rep) {
+            _Expected_rep->_Decref();
+        }
+        return false;
+    }
+
+    constexpr atomic() noexcept = default;
+
+    atomic(const shared_ptr<_Ty> _Value) noexcept : _Base(_Value._Ptr, _Value._Rep) {
+        _Value._Incref();
+    }
+
+    atomic(const atomic&) = delete;
+    void operator=(const atomic&) = delete;
+
+    void operator=(shared_ptr<_Ty> _Value) noexcept {
+        store(_STD move(_Value));
+    }
+
+    ~atomic() {
+        const auto _Rep = reinterpret_cast<_Ref_count_base*>(this->_Repptr.load(memory_order::relaxed));

```

Comment:
Just checking, this can be weaker than SC because destruction shouldn't be simultaneous with anything else, right?

Reply:
Yes